### PR TITLE
(PUP-9395) specify version-specific documentation parameters during install

### PR DESCRIPTION
**Pull request by tkishel**
With this commit, install specifies version-specific options for documentation.
Apply those options independent of the :source parameter.

### CODE REVIEW

**Comment by joshcooper** [comment]
@tkishel could you rebase your PR on the latest 4.10.x (now that the bundler 2 fix is in)?

File: lib/puppet/provider/package/gem.rb [hunk 1]
```diff
@@ -121,13 +127,18 @@ def install(useversion = true)
           command << source
         else
           # interpret it as a gem repository
-          command << "--source" << "#{source}" << resource[:name]
+          command << "--source" << "#{source}"
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: This looks unrelated to the `--no-document` issue?

File: lib/puppet/provider/package/gem.rb [hunk 2]
```diff
@@ -90,8 +90,14 @@ def insync?(is)
     is.any? { |version| dependency.match?('', version) }
   end
 
+  def gem_version(cmd)
+    command = [cmd, '--version']
+    execute(command, {:failonfail => true, :combine => true, :custom_environment => {"HOME"=>ENV["HOME"]}})
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: I didn't follow this part, why pass `failonfail`, etc?

File: lib/puppet/provider/package/gem.rb [hunk 3]
```diff
@@ -121,13 +127,18 @@ def install(useversion = true)
           command << source
         else
           # interpret it as a gem repository
-          command << "--source" << "#{source}" << resource[:name]
+          command << "--source" << "#{source}"
         end
       end
+    end
+
+    if Puppet::Util::Package.versioncmp(gem_version(command(:gemcmd)), '2.0.0') == -1
```
Comment (joshcooper) [inline lib/puppet/provider/package/gem.rb]: Does `versioncmp` always return `-1`, `0`, `1` or can it return a more negative version? Guess I'm more used to seeing `versioncmp(a, b) < 0`, or `== 0`, or `> 0`

File: lib/puppet/provider/package/gem.rb (hunk 1 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: The `if source = resource[:source]` statement should be limited to adding `source` to the `command` array. We should always add `--no-rdoc ...` to the `command` array to never install documentation: independent of the whether or not a :source parameter is defined. They are absolutely unrelated.

File: lib/puppet/provider/package/gem.rb (hunk 2 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Sorry, that comes from newer branches, like:

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/package/gem.rb#L134

I will revert to use `execute(command)` to match this branch.

File: lib/puppet/provider/package/gem.rb (hunk 3 above)
Comment (tkishel) [reply lib/puppet/provider/package/gem.rb]: Yes: 

https://github.com/puppetlabs/puppet/blob/4.10.x/lib/puppet/functions/versioncmp.rb#L15

Chose that to match this: 

https://github.com/puppetlabs/puppet/blob/4.10.x/lib/puppet/provider/package/pip.rb#L81

**Comment by branan** [comment]
I don't think this needs to go into 4.10.x? 5.5 seems fine to me. Landing any code in 4.10.x will cause a CI run, which is just extraneous load if we're never going to need to ship this

**Comment by joshcooper** [comment]
Superseded by PR #7320
